toolkit/tools/imager/imager.go

```diff
@@ -328,50 +335,81 @@ func setupRootFS(outputDir, installRoot string) (extraMountPoints []*safechroot.
 	const (
 		realDiskType = "path"
 	)
-	if diskConfig.TargetDisk.Type == realDiskType {
+	var (
+		DiskType = realDiskType
+	)
+
+	logger.Log.Infof("configs (%d)", len(disks))
+
+	for i, disk := range disks {
+		if i == 0 {
+			DiskType = disk.TargetDisk.Type
+			continue
+		}
+		if disk.TargetDisk.Type != DiskType {
+			err = fmt.Errorf("Only one disk type can be specified found (%s) and (%s)", DiskType, disk.TargetDisk.Type)
+			return
+		}
+
+	}
+
+	if DiskType == realDiskType {
 		if liveInstallFlag {
-			diskDevPath = diskConfig.TargetDisk.Value
-			partIDToDevPathMap, partIDToFsTypeMap, encryptedRoot, readOnlyRoot, err = setupRealDisk(diskDevPath, diskConfig, rootEncryption, readOnlyRootConfig)
+			partIDToDevPathMap, partIDToFsTypeMap, encryptedRoot, readOnlyRoot, err = setupRealDisks(diskNames, disks, rootEncryption, readOnlyRootConfig)
 		} else {
 			err = fmt.Errorf("target Disk Type is set but --live-install option is not set. Please check your config or enable the --live-install option")
 			return
 		}
 	} else {
-		diskDevPath, partIDToDevPathMap, partIDToFsTypeMap, encryptedRoot, readOnlyRoot, err = setupLoopDeviceDisk(outputDir, diskName, diskConfig, rootEncryption, readOnlyRootConfig)
+		diskDevPaths, partIDToDevPathMap, partIDToFsTypeMap, encryptedRoot, readOnlyRoot, err = setupLoopDeviceDisks(outputDir, diskNames, disks, rootEncryption, readOnlyRootConfig)
 		isLoopDevice = true
 	}
 	return
 }
 
-func setupLoopDeviceDisk(outputDir, diskName string, diskConfig configuration.Disk, rootEncryption configuration.RootEncryption, readOnlyRootConfig configuration.ReadOnlyVerityRoot) (diskDevPath string, partIDToDevPathMap, partIDToFsTypeMap map[string]string, encryptedRoot diskutils.EncryptedRootDevice, readOnlyRoot diskutils.VerityDevice, err error) {
+func setupLoopDeviceDisks(outputDir string, diskNames []string, diskConfigs []configuration.Disk, rootEncryption configuration.RootEncryption, readOnlyRootConfig configuration.ReadOnlyVerityRoot) (diskDevPaths []string, partIDToDevPathMap, partIDToFsTypeMap map[string]string, encryptedRoot diskutils.EncryptedRootDevice, readOnlyRoot diskutils.VerityDevice, err error) {
 	defer func() {
 		// Detach the loopback device on failure
-		if err != nil && diskDevPath != "" {
-			detachErr := diskutils.DetachLoopbackDevice(diskDevPath)
-			if detachErr != nil {
-				logger.Log.Errorf("Failed to detach loopback device on failed initialization. Error: %s", detachErr)
+		if err != nil && len(diskDevPaths) != 0 {
+			for _, diskDevPath := range diskDevPaths {
+				detachErr := diskutils.DetachLoopbackDevice(diskDevPath)
```

Comment:
Any chance we could have a list of paths that are `""` here? Might be worth still checking that inside the loop.

Reply:
I dont think it is possible for the diskdevpaths to be "", they will all be the value "{some_path}/disk#.raw" which is derived from the DefaultTempDiskNames which will always be defined. I am not super familiar with golang so would diskdevpaths be modified before running the defer functioned? If so then i should make the change since some other function at some point could add "" to the list. I think If diskdevpaths contained a "" then that would be caught when calling SetupLoopbackDevice()

Reply:
Yes, parameters to a defer call are passed at the time the call is executed (mutable). It seems likely that "" won't get added, so probably fine.